tests/test_mutations.py

```diff
@@ -472,40 +704,226 @@ def test_keep_mutation_parent_zero_rate(self):
+            for m in site_item['mutation_list']:
+                mut = m['mutation']
+                parent = mut.parent
+                if parent != tskit.NULL:
+                    parent += new_mutations
+                tables.mutations.add_row(site_id, mut.node, mut.derived_state,
+                                         parent=parent, metadata=mut.metadata)
+                if m['new']:
+                    new_mutations += 1
+
+    def place_mutations(self, tables, rng, discrete_sites=False):
+        # Insert a sentinel into the map for convenience.
+        map_position = np.hstack([self.rate_map.position, [tables.sequence_length]])
+        node_times = tables.nodes.time
+        for edge in tables.edges:
+            branch_start = node_times[edge.child]
+            branch_end = node_times[edge.parent]
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.value[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    time = rng.flat(branch_start, branch_end)
+                    site_pos = np.searchsorted([s['site'].position for s in self.sites],
+                                               position)
+                    if (site_pos == len(self.sites)
+                       or self.sites[site_pos]['site'].position != position):
+                        # new site
+                        ancestral_state = self.alleles[self.pick_allele()]
+                        site = tskit.SiteTableRow(position=position,
+                                                  ancestral_state=ancestral_state,
+                                                  metadata=b'')
+                        site_item = {'site': site,
+                                     'mutation_list': []}
+                        self.sites.insert(site_pos, site_item)
+                    else:
+                        site_item = self.sites[site_pos]
+                    mut_pos = np.searchsorted([-1 * m['time']
```

Comment:
This is obscure - what are you doing here?

Reply:
keeping mutations reverse-sorted by time - is there a better python idiom for this?